crates/consensus/src/transaction/eip4844/builder.rs

```diff
@@ -0,0 +1,408 @@
+#[cfg(not(feature = "kzg"))]
+use alloy_eips::eip4844::Blob;
+#[cfg(feature = "kzg")]
+use c_kzg::{Blob, KzgCommitment, KzgProof, KzgSettings};
+
+use alloy_eips::eip4844::{BYTES_PER_BLOB, FIELD_ELEMENTS_PER_BLOB};
+
+use super::utils::WholeFe;
+
+/// A builder for creating a [`BlobTransactionSidecar`].
+///
+/// [`BlobTransactionSidecar`]: crate::BlobTransactionSidecar
+#[derive(Debug, Clone)]
+pub struct PartialSidecar {
+    /// The blobs in the sidecar.
+    blobs: Vec<Blob>,
+    /// The number of field elements that we have ingested, total.
+    fe: usize,
+}
+
+impl Default for PartialSidecar {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PartialSidecar {
+    /// Create a new builder.
+    pub fn new() -> Self {
+        // NB: vecs default to 100 capacity. Blobs are large. We don't want
+        // to allocate 100 blobs if we don't need them.
+        Self::with_capacity(2)
+    }
+
+    /// Create a new builder with a given capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut blobs = Vec::with_capacity(capacity);
+        blobs.push(Blob::new([0u8; BYTES_PER_BLOB]));
+        Self { blobs, fe: 0 }
+    }
+
+    /// Get a reference to the blobs currently in the builder.
+    pub fn blobs(&self) -> &[Blob] {
+        &self.blobs
+    }
+
+    /// Get the number of unused field elements that have been allocated
+    fn free_fe(&self) -> usize {
+        self.blobs.len() * FIELD_ELEMENTS_PER_BLOB as usize - self.fe
+    }
+
+    /// Calculate the length of used field elements IN BYTES in the builder.
+    ///
+    /// This is always strictly greater than the number of bytes that have been
+    /// ingested.
+    pub const fn len(&self) -> usize {
+        self.fe * 32
+    }
+
+    /// Check if the builder is empty.
+    pub const fn is_empty(&self) -> bool {
+        self.fe == 0
+    }
+
+    /// Push an empty blob to the builder, and reset the unused counter.
+    fn push_empty_blob(&mut self) {
+        self.blobs.push(Blob::new([0u8; BYTES_PER_BLOB]));
+    }
+
+    /// Allocate enough space for the required number of new field elements.
+    pub fn alloc_fes(&mut self, required_fe: usize) {
+        while self.free_fe() < required_fe {
+            self.push_empty_blob()
+        }
+    }
+
+    /// Get the number of used field elements in the current blob.
+    const fn fe_in_current_blob(&self) -> usize {
+        self.fe % FIELD_ELEMENTS_PER_BLOB as usize
+    }
+
+    /// Get the index of the first unused field element in the current blob.
+    const fn first_unused_fe_index_in_current_blob(&self) -> Option<usize> {
+        if self.fe_in_current_blob() as u64 == FIELD_ELEMENTS_PER_BLOB {
```

Comment:
is this ever the case? since the method is `self.fe % FIELD_ELEMENTS_PER_BLOB`

Reply:
correct. this was leftover from an earlier design. removing